syz-fuzzer/fuzzer.go

```diff
@@ -514,28 +514,67 @@ func (fuzzer *Fuzzer) checkDisabledCalls(p *prog.Prog) {
 	}
 }
 
-func (fuzzer *FuzzerSnapshot) chooseProgram(r *rand.Rand) *prog.Prog {
-	randVal := r.Int63n(fuzzer.sumPrios + 1)
-	idx := sort.Search(len(fuzzer.corpusPrios), func(i int) bool {
-		return fuzzer.corpusPrios[i] >= randVal
-	})
-	return fuzzer.corpus[idx]
+func (fuzzer *Fuzzer) updateCorpusMap(hashes signal.Signal, p *prog.Prog) {
+	for hashRaw := range hashes.Serialize().Elems {
+		hash := uint32(hashRaw)
+		old, ok := fuzzer.perHashProgs[hash]
+		if !ok {
+			fuzzer.allHashes = append(fuzzer.allHashes, hash)
+		}
+		// If a hash is covered by too many progs already, it does not matter
+		// anymore by how many exactly.
+		const perHashLimit = 25
+		if len(old) > perHashLimit {
+			continue
+		}
+		fuzzer.perHashProgs[hash] = append(old, p)
+	}
+}
+
+func (fuzzer *Fuzzer) chooseProgram(r *rand.Rand) *prog.Prog {
```

Comment:
One forgotten case (that sometimes fails the test) is when there's no signal.

Store some separate slice with such programs and pick a random one if `allHashes` is empty?